File: hilos1.py
#Este programa muestra una simulacion senccilla de como funcionan los hilos
import threading  # Importa el módulo threading para usar hilos
import time       # Importa el módulo time para manejar pausas en la ejecución

# Definimos una función que será ejecutada por los hilos
def imprimir_mensaje(nombre, veces):  # La función recibe un nombre y cuántas veces ejecutar el bucle
    for i in range(veces):            # Repite el bucle `n` veces
        print(f"{nombre}: Ejecutando iteración {i + 1}")  # Imprime el nombre del hilo y la iteración actual
        time.sleep(1)  # Pausa la ejecución por 1 segundo para simular una tarea lenta

# Inicia el contador de tiempo
inicio = time.time()

# Crear los hilos
hilo1 = threading.Thread(target=imprimir_mensaje, args=("Hilo 1", 3))  # Asocia la función y sus argumentos al hilo
hilo2 = threading.Thread(target=imprimir_mensaje, args=("Hilo 2", 2))  # Segundo hilo con otros argumentos

# Iniciar los hilos
hilo1.start()  # Comienza la ejecución del hilo 1
hilo2.start()  # Comienza la ejecución del hilo 2

# Esperar a que los hilos terminen
hilo1.join()  # Bloquea la ejecución principal hasta que el hilo 1 termine
hilo2.join()  # Bloquea la ejecución principal hasta que el hilo 2 termine

# Detiene el contador de tiempo
fin = time.time()

# Calcula el tiempo total
tiempo_total = fin - inicio

print("Todos los hilos han terminado.")  # Imprime un mensaje final cuando ambos hilos terminan
print(f"Tiempo total: {tiempo_total:.2f} segundos.")  # Imprime el tiempo total

File: hilos1.java
package Java;
// Este programa muestra una simulación sencilla de cómo funcionan los hilos en Java

class Hilo implements Runnable {
    private String nombre;
    private int veces;

    // Constructor para inicializar el nombre del hilo y el número de iteraciones
    public Hilo(String nombre, int veces) {
        this.nombre = nombre;
        this.veces = veces;
    }

    @Override
    public void run() {
        for (int i = 0; i < veces; i++) {
            System.out.println(nombre + ": Ejecutando iteración " + (i + 1));
            try {
                Thread.sleep(1000); // Pausa la ejecución por 1 segundo
            } catch (InterruptedException e) {
                System.out.println(nombre + " fue interrumpido.");
            }
        }
    }
}

public class hilos1 {
    public static void main(String[] args) {
        // Inicia el contador de tiempo
        long inicio = System.currentTimeMillis();

        // Crear los hilos
        Thread hilo1 = new Thread(new Hilo("Hilo 1", 3));
        Thread hilo2 = new Thread(new Hilo("Hilo 2", 2));

        // Iniciar los hilos
        hilo1.start();
        hilo2.start();

        // Esperar a que los hilos terminen
        try {
            hilo1.join();
            hilo2.join();
        } catch (InterruptedException e) {
            System.out.println("La ejecución principal fue interrumpida.");
        }

        // Detiene el contador de tiempo
        long fin = System.currentTimeMillis();

        // Calcula el tiempo total
        double tiempoTotal = (fin - inicio) / 1000.0;

        System.out.println("Todos los hilos han terminado.");
        System.out.printf("Tiempo total: %.2f segundos.%n", tiempoTotal);
    }
}
